src/main.rs

```diff
@@ -1311,78 +1311,100 @@ fn parse_manpage_at_path(
-//             add_diagnostic('Cannot open ' + manpage_path)
-//         except (KeyboardInterrupt, SystemExit):
-//             raise
-//         except:
-//             add_diagnostic('Error parsing %s: %s' % (manpage_path, sys.exc_info()[0]), BRIEF_VERBOSE)
-//             flush_diagnostics(sys.stderr)
-//             traceback.print_exc(file=sys.stderr)
-//         flush_diagnostics(sys.stderr)
-//     print("") #Newline after loop
-//     add_diagnostic("Successfully parsed %d / %d pages" % (successful_count, total_count), BRIEF_VERBOSE)
-//     flush_diagnostics(sys.stderr)
-
-#[derive(Copy, Clone, Debug)]
-struct Progress(pub bool);
-
-// TODO Arg/output types?
+/// Get the number of digits in num
+fn num_digits(n: usize) -> usize {
+    (1.max(n) as f32).log10() as usize + 1
+}
+
+#[test]
+fn test_num_digits() {
+    assert_eq!(num_digits(1000), 4);
+    assert_eq!(num_digits(100), 3);
+    assert_eq!(num_digits(33), 2);
+    assert_eq!(num_digits(123456789012345), 15);
+    assert_eq!(num_digits(0), 1);
+}
+
 fn parse_and_output_man_pages(
-    _paths: impl Iterator<Item = PathBuf>,
-    _output_directory: PathBuf,
-    Progress(_show_progress): Progress,
+    paths: &mut [PathBuf],
+    output_directory: PathBuf,
+    show_progress: bool,
+    deroff_only: bool,
+    write_to_stdout: bool,
 ) {
-    unimplemented!();
+    paths.sort();
+
+    let total = paths.len();
+
+    let mut successful_count = 0;
+    let padding = num_digits(total);
+
+    let mut last_len: usize = 0;
+
+    if show_progress && !write_to_stdout {
+        println!(
+            "Parsing man pages and writing completions to {:?}",
+            output_directory
+        );
+    }
+
+    // As far as I can tell, we're fully supporting xz / lzma, so the checks here aren't needed
+
+    for (index, manpage_path) in paths.iter().enumerate() {
+        // foo/bar/gcc.1.gz -> gcc.1.gz
+        // TODO: Expect
+        let man_file_name = manpage_path
+            .file_name()
+            .map(|fname| fname.to_string_lossy())
+            .unwrap_or_else(|| panic!("Failed to get manfile name from {:?}", manpage_path));
+
+        // gcc.1.gz -> gcc
+        // `str::split` iterator ALWAYS has a first element by definition
+        let cmd_name = man_file_name.split('.').next().unwrap();
+
+        if show_progress && !write_to_stdout {
+            let progress = format!(
+                "  {0:>1$} / {2} : {3}",
+                index + 1,
+                padding,
+                total,
+                man_file_name,
+            );
+
+            // padded_progress_str = progress_str.ljust(last_progress_string_length)
+            let padded = format!("{0:<1$}", progress, last_len);
+
+            last_len = progress.len();
+            // TODO: Expects
+            let stdout = std::io::stdout();
+            let mut lock = stdout.lock();
+            lock.write_all(format!("\r{}\r", padded).as_bytes())
+                .expect("Failed to write to stdout");
```

Comment:
I don't think we need to be so old rigid. Why not just clear current line so we don't need to keep track of how many items we have printed? `\e[K{}\r`, `\e[K` clear current line from cursor to end, we could remove `last_len` if we do this.

Reply:
I'm not familiar with `\e[K\r`?

Reply:
Just change that to `\e[K{}\r`. It should work, I tested. Try on your terminal?
```
$ echo \e '\e[Khello world\r\e[Kyay\r'
```
It should show `yay`.

Reply:
Looks like Rust isn't familiar with `\e` as an escape sequence.

Reply:
@pickfire [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6d1d2e2c995894962aae994f4c81bc9e)

Reply:
Ah, you need to unescape that in rust, either use `r"\e[K{}\r"` or `"\\e[K{}\\r"`, I prefer the first.

Reply:
Am I missing something?
```rust
// Contents of test.rs
fn main() { 
    for i in 0..10 {
        println!(r"\e[K{}\r", i);
            
    }
}
```
```
c@flesh ~/g/fish-manpage-completions (deroff)> nv test.rs
c@flesh ~/g/fish-manpage-completions (deroff)> rustc test.rs 
c@flesh ~/g/fish-manpage-completions (deroff)> ./test
\e[K0\r
\e[K1\r
\e[K2\r
\e[K3\r
\e[K4\r
\e[K5\r
\e[K6\r
\e[K7\r
\e[K8\r
\e[K9\r
```

Reply:
We could just `// TODO: pickfire thinks this is ugly` this

Reply:
```rust
> cat > src/main.rs
use std::io::{self, Write};
use std::thread;
use std::time::Duration;

fn main() {
    for i in 0..10 {
        let stdout = io::stdout();
        let mut stdout = stdout.lock();
        stdout.write_all(format!("\r\x1b[K{}", i).as_bytes()).ok();
        stdout.flush().ok();
        thread::sleep(Duration::from_secs(1));
    }
    println!();
}
> rustc main.rs
> ./main.rs
9
```

Reply:
10a0120